### PR TITLE
Hotfix/nil middleware browser monitoring

### PR DESCRIPTION
**Pull request by mwlang**
# Overview
Attempts to resolve scenario where BrowserMonitoring middleware gets added to the middleware stack multiple times.
Sinatra doesn't initial the middle stack until the first request is processed.  It appears from the issue described #563 that the agent's code for adding to the middleware stack is not thread-safe.  However, this is not easily reproducible, so starting this PR in draft and looking for confirmation of fix from the user community.

# Related Github Issue
* #563

### CODE REVIEW

**Comment by mwlang** [comment]
This PR may be tried while unmerged by adding the following to your `Gemfile`

```Ruby
gem 'newrelic_rpm', 
  :git => 'https://github.com/newrelic/newrelic-ruby-agent.git',  
  :branch => 'hotfix/nil_middleware_browser_monitoring'
```

**Comment by alexJunger** [comment]
Thanks @mwlang, this looks promising! The cause you identified would also explain why it's not a common issue.
Will give it a try today!

**Comment by Goltergaul** [comment]
@mwlang very good news, thanks a lot :partying_face: We'll try this asap

**Comment by mwlang** [comment]
I've added a guard against `app.middleware` being nil.  This seems to be happening because app spins up and starts shutting down in rapid succession and the Ruby agent is going through the motion of instrumenting the middleware stack in this moment.  Purely a guess!  However, this fix will at least prevent start up errors from the Agent getting in the way during such occurrences, thus allowing the application to run unhindered.

```
undefined method `[]' for nil:NilClass

/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/sinatra.rb:110:in `block in try_to_use'
org/jruby/RubyArray.java:4493:in `any?'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/sinatra.rb:110:in `block in try_to_use'
org/jruby/ext/thread/Mutex.java:164:in `synchronize'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/sinatra.rb:109:in `try_to_use'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/sinatra.rb:98:in `block in build_with_newrelic'
org/jruby/RubyArray.java:1809:in `each'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/sinatra.rb:97:in `build_with_newrelic'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/sinatra-2.1.0/lib/sinatra/base.rb:1528:in `new'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/sinatra-2.1.0/lib/sinatra/base.rb:1517:in `prototype'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/sinatra-2.1.0/lib/sinatra/base.rb:1542:in `block in call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/sinatra-2.1.0/lib/sinatra/base.rb:1769:in `synchronize'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/sinatra-2.1.0/lib/sinatra/base.rb:1542:in `call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/rack-2.2.3/lib/rack/urlmap.rb:74:in `block in call'
org/jruby/RubyArray.java:1809:in `each'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/rack-2.2.3/lib/rack/urlmap.rb:58:in `call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/bundler/gems/newrelic-ruby-agent-fd7341174889/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/configuration.rb:246:in `call'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/request.rb:76:in `block in handle_request'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/thread_pool.rb:337:in `with_force_shutdown'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/request.rb:75:in `handle_request'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/server.rb:431:in `process_client'
/opt/app-root/service/vendor/bundle/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/thread_pool.rb:145:in `block in spawn_thread'
```

**Comment by Goltergaul** [comment]
@mwlang Thank you for the update, I'm rolling this out and will report back via the support ticket after running it for a while

**Comment by namedpipe** [comment]
We've deployed 2d451e3 to our production system that is using jruby. While the error rate is less, we still do get occasional errors.

Here's the last error we got this morning:

```
NoMethodError: undefined method `[]' for nil:NilClass
…e3e17c6/lib/new_relic/agent/instrumentation/sinatra.rb:  110:in `block in try_to_use'
org/jruby/RubyArray.java:4493:in `any?'
…e3e17c6/lib/new_relic/agent/instrumentation/sinatra.rb:  110:in `block in try_to_use'
org/jruby/ext/thread/Mutex.java:164:in `synchronize'
…e3e17c6/lib/new_relic/agent/instrumentation/sinatra.rb:  109:in `try_to_use'
…e3e17c6/lib/new_relic/agent/instrumentation/sinatra.rb:   98:in `block in build_with_newrelic'
org/jruby/RubyArray.java:1809:in `each'
…e3e17c6/lib/new_relic/agent/instrumentation/sinatra.rb:   97:in `build_with_newrelic'
…dle/jruby/2.5.0/gems/sinatra-1.4.8/lib/sinatra/base.rb: 1473:in `new'
…dle/jruby/2.5.0/gems/sinatra-1.4.8/lib/sinatra/base.rb: 1462:in `prototype'
…dle/jruby/2.5.0/gems/sinatra-1.4.8/lib/sinatra/base.rb: 1487:in `block in call'
…dle/jruby/2.5.0/gems/sinatra-1.4.8/lib/sinatra/base.rb: 1787:in `synchronize'
…dle/jruby/2.5.0/gems/sinatra-1.4.8/lib/sinatra/base.rb: 1487:in `call'
…/new_relic/agent/instrumentation/middleware_tracing.rb:  101:in `call'
…/gems/rack-ssl-enforcer-0.2.9/lib/rack/ssl-enforcer.rb:   52:in `call'
…/new_relic/agent/instrumentation/middleware_tracing.rb:  101:in `call'
…gems/sentry-raven-2.7.4/lib/raven/integrations/rack.rb:   51:in `call'
…/new_relic/agent/instrumentation/middleware_tracing.rb:  101:in `call'
…y/2.5.0/gems/puma-5.1.1-java/lib/puma/configuration.rb:  246:in `call'
…e/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/request.rb:   76:in `block in handle_request'
…uby/2.5.0/gems/puma-5.1.1-java/lib/puma/thread_pool.rb:  337:in `with_force_shutdown'
…e/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/request.rb:   75:in `handle_request'
…le/jruby/2.5.0/gems/puma-5.1.1-java/lib/puma/server.rb:  431:in `process_client'
…uby/2.5.0/gems/puma-5.1.1-java/lib/puma/thread_pool.rb:  145:in `block in spawn_thread'
```

We can try deploying 4da7f57 next week to see if all the errors go away.

**Comment by Goltergaul** [comment]
https://github.com/newrelic/newrelic-ruby-agent/commit/4da7f576987690e36588def86398416a808c8ad7 is looking promising so far on our systems. I'm still keeping an eye on it

**Comment by angelatan2** [comment]
@Goltergaul @namedpipe Thanks for sharing the details of your tests.  We are looking forward to hearing the results and or error logs if it fails.

**Comment by Goltergaul** [comment]
I did one more round of testing and was not able to reproduce it so far. With the other versions where i could reproduce it i was running into it rather fast but now it looks good so far :crossed_fingers:

**Comment by mwlang** [comment]
@Goltergaul Thank you for testing this out and sharing your results.  This was a really tricky blind bug fix and our team is really thankful for the consistent feedback along the way.

**Comment by namedpipe** [comment]
We put 4da7f57 into production yesterday. But we've only been seeing the error about once a week when we scale up. That usually happens on Fridays. So we'll know by the end of the week how it's panning out.

**Comment by mwlang** [comment]
@namedpipe sounds great and thanks for the update.  If this is resolved for you as well, then I think we can confidently call it a fix.

**Comment by namedpipe** [comment]
We didn't see any errors last week. Feels like this issue is gone now.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,14 @@
 # New Relic Ruby Agent Release Notes #
 
+  ## v6.15.1
```
Comment (tannalynn) [inline CHANGELOG.md:3]: Are we planning on doing a 6.15.1 release? Or should this be 7.0.0?

Comment (Goltergaul) [reply CHANGELOG.md:3]: if it would be feasible it would be nice to have the fix sooner :)

Comment (mwlang) [reply CHANGELOG.md:3]: I wasn't sure where our timeline was falling for 7.0 at the time I wrote 6.15.1.  

7.0.0 release is imminent, so I think we'll skip releasing a 6.15.1

**Review by tannalynn — APPROVED** [review]
Looks great!

**Pull request merged**